SIM-ONE Training/simone_transformer/modern_layers.py

```
@@ -83,111 +83,159 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
+        topk_logits, topk_indices = torch.topk(router_logits, k, dim=-1)
+        topk_weights = F.softmax(topk_logits, dim=-1)
+
+        # Flatten routing information once for vectorized dispatch
+        token_indices = torch.arange(num_tokens, device=x_flat.device)
+        expanded_tokens = token_indices.unsqueeze(-1).expand(-1, k).reshape(-1)
+        expanded_experts = topk_indices.reshape(-1)
+        expanded_weights = topk_weights.reshape(-1)
+
+        # Capacity-aware routing (Drop lowest scores when an expert is overloaded)
+        if self.capacity_factor is not None and self.capacity_factor > 0:
+            capacity = max(1, int(math.ceil(self.capacity_factor * num_tokens / self.num_experts)))
```

Ruff (0.13.1)

Value being cast to `int` is already an integer

Remove unnecessary `int` call

(RUF046)